pkg/scheduler/framework/statement.go

```diff
@@ -194,7 +196,117 @@ func (s *Statement) unpipeline(task *api.TaskInfo) error {
 	return nil
 }
 
-// Discard operation for evict and pipeline
+// Save the allocate operation on task
+func (s *Statement) Allocate(task *api.TaskInfo, hostname string) error {
+	if err := s.ssn.cache.AllocateVolumes(task, hostname); err != nil {
+		return err
+	}
+
+	// Only update status in session
+	job, found := s.ssn.Jobs[task.Job]
+	if found {
+		if err := job.UpdateTaskStatus(task, api.Allocated); err != nil {
+			glog.Errorf("Failed to update task <%v/%v> status to %v in Session <%v>: %v",
+				task.Namespace, task.Name, api.Allocated, s.ssn.UID, err)
+			return err
+		}
+	} else {
+		glog.Errorf("Failed to found Job <%s> in Session <%s> index when binding.",
+			task.Job, s.ssn.UID)
+		return fmt.Errorf("failed to find job %s", task.Job)
+	}
+
+	task.NodeName = hostname
+
+	if node, found := s.ssn.Nodes[hostname]; found {
+		if err := node.AddTask(task); err != nil {
+			glog.Errorf("Failed to add task <%v/%v> to node <%v> in Session <%v>: %v",
+				task.Namespace, task.Name, hostname, s.ssn.UID, err)
+			return err
+		}
+		glog.V(3).Infof("After allocated Task <%v/%v> to Node <%v>: idle <%v>, used <%v>, releasing <%v>",
+			task.Namespace, task.Name, node.Name, node.Idle, node.Used, node.Releasing)
+	} else {
+		glog.Errorf("Failed to found Node <%s> in Session <%s> index when binding.",
+			hostname, s.ssn.UID)
+		return fmt.Errorf("failed to find node %s", hostname)
+	}
+
+	// Callbacks
+	for _, eh := range s.ssn.eventHandlers {
+		if eh.AllocateFunc != nil {
+			eh.AllocateFunc(&Event{
+				Task: task,
+			})
+		}
+	}
+
+	// Update status in session
+	glog.V(3).Info("Allocating operations ...")
+	s.operations = append(s.operations, operation{
+		name: "allocate",
+		args: []interface{}{task, hostname},
+	})
+
+	return nil
+}
+
+func (s *Statement) allocate(task *api.TaskInfo, hostname string) error {
```

Comment:
Is it same as `ssn.dispatch()`? Could we reuse it?

Reply:
Great comments. The answer is currently the allocation func in statement is an enhancement compared to the allocation.go file.  Disscussed with Klaus, the principle is we will only use the Allocate  func in statement. The func in allocte.go will be deprecated in the future.